libvast/src/system/transformer.cpp

```diff
@@ -55,31 +55,68 @@ transformer_stream_stage_ptr attach_transform_stage(
-  VAST_TRACE_SCOPE("{}", VAST_ARG(name));
+  VAST_TRACE_SCOPE("transformer {} {}", VAST_ARG(self->id()), VAST_ARG(name));
   self->state.transformer_name = std::move(name);
   auto transform_names = list{};
   for (const auto& t : transforms)
     transform_names.emplace_back(t.name());
+  self->state.source_requires_shutdown = false;
   self->state.status["transforms"] = std::move(transform_names);
   self->state.transforms = transformation_engine{std::move(transforms)};
   self->state.stage = attach_transform_stage(self);
   return {
     [self](const stream_sink_actor<table_slice>& out)
       -> caf::outbound_stream_slot<table_slice> {
+      if (out->getf(caf::abstract_actor::is_shutting_down_flag)) {
+        VAST_DEBUG("{} ignores stream sink {} because the actor is already "
+                   "shutting down",
+                   self->state.transformer_name, out);
+        return {};
+      }
       VAST_DEBUG("{} adds stream sink {}", self->state.transformer_name, out);
       return self->state.stage->add_outbound_path(out);
     },
     [self](const stream_sink_actor<table_slice, std::string>& sink,
            std::string name) {
+      if (sink->getf(caf::abstract_actor::is_shutting_down_flag)) {
+        VAST_DEBUG("{} ignores stream sink {} because the actor is already "
+                   "shutting down",
+                   self->state.transformer_name, sink);
+        return;
+      }
       VAST_DEBUG("{} adds stream sink {} (registers as {})",
                  self->state.transformer_name, sink, name);
       self->state.stage->add_outbound_path(sink,
                                            std::make_tuple(std::move(name)));
     },
     [self](caf::stream<detail::framed<table_slice>> in)
       -> caf::inbound_stream_slot<detail::framed<table_slice>> {
+      // There's a race condition (mostly when using `vast -N`) that prevents
+      // shutdown when the importer shuts down before the stream handshake
+      // finishes. In this case the `eof` it sends never arrives (because we
+      // call `push()` manually and there are no paths at that time) and thus
+      // the transformer never quits, so we have to exit manually here. We also
+      // have to open a stream slot and `shutdown()` the stream again, or the
+      // importer will be stuck with a pending stream manager and not be
+      // properly cleaned up, also causing a deadlock on shutdown. On the other
+      // hand, some sources will shut down independently of the system shutdown,
+      // e.g. when importing from a file. In this case we may have the opposite
+      // race, where there is still data that the stream manager needs to push
+      // onto the stream, and we must not shutdown in this case since that would
+      // cause the removal of the path upstream. Since we cannot possibly tell
+      // from here which behavior is correct, we require the party that spawns
+      // the transformer to decide what they require.
```

Comment:
Note: I tried to make this less of a wall of text by adding some paragraphs, but `clang-format` doesn't like them :/